src/hdmf/common/resources.py

```diff
@@ -0,0 +1,355 @@
+import uuid
+import numpy as np
+import pandas as pd
+
+from . import register_class
+from ..container import Table, Row, Container
+
+from ..utils import docval, popargs
+
+
+def _check_id(table, id):
+    if id >= 0 and len(table.which(id=id)) == 0:
+        return np.uint64(id)
+    else:
+        raise ValueError('id must be a non-negative integer that is not already in the table: %d' % id)
+
+
+class KeyTable(Table):
+    """
+    A table for storing keys used to reference external resources
+    """
+
+    __defaultname__ = 'keys'
+
+    __columns__ = (
+        {'name': 'key_name', 'type': str,
+         'doc': 'The user key that maps to the resource term / registry symbol.'},
+    )
+
+
+class Key(Row):
+    """
+    A Row class for representing rows in the KeyTable
+    """
+
+    __table__ = KeyTable
+
+
+class ResourceTable(Table):
+    """
+    A table for storing the external resources a key refers to
+    """
+
+    __defaultname__ = 'resources'
+
+    __columns__ = (
+        {'name': 'keytable_idx', 'type': (int, str, Key),
+         'doc': 'The user key that maps to the resource term / registry symbol.'},
+        {'name': 'resource_name', 'type': str,
+         'doc': 'The resource/registry that the term/symbol comes from.'},
+        {'name': 'resource_entity_id', 'type': str,
+         'doc': 'The unique resource identifier for the resource term / registry symbol.'},
+        {'name': 'resource_entity_uri', 'type': str,
+         'doc': 'The unique resource identifier for the resource term / registry symbol.'},
+    )
+
+
+class ResourceEntity(Row):
+    """
+    A Row class for representing rows in the ResourceTable
+    """
+
+    __table__ = ResourceTable
+
+
+class ObjectTable(Table):
+    """
+    A table for storing objects that contain keys that refer to external resources
+    """
+
+    __defaultname__ = 'objects'
+
+    __columns__ = (
+        {'name': 'object_id', 'type': (str, uuid.UUID),
```

Comment:
Is there code to convert the uuid.UUID into a string somewhere, and is this tested? It might be easier to allow only strings.

Reply:
AbstractContainer.object_id returns a string anyway, so I am not sure when a user would have a `uuid.UUID` object